plugins/flat-object/src/main/java/org/opensearch/flatobject/mapper/FlatObjectFieldMapper.java

```diff
@@ -8,111 +8,537 @@
+            fieldtype.setStored(this.stored.getValue());
+            return new FlatObjectFieldMapper(
+                name,
+                fieldtype,
+                buildFieldType(context, fieldtype),
+                multiFieldsBuilder.build(this, context),
+                copyTo.build(),
+                this
+            );
         }
     }
 
-    public static class TypeParser implements Mapper.TypeParser {
+    public static final TypeParser PARSER = new TypeParser((n, c) -> new Builder(n, c.getIndexAnalyzers()));
+
+    /**
+     * Field type for flat-object fields
+     *
+     * @opensearch.internal
+     */
+    public static final class FlatObjectFieldType extends StringFieldType {
+
+        private final int ignoreAbove;
+        private final String nullValue;
+
+        public FlatObjectFieldType(
+            String name,
+            FieldType fieldType,
+            NamedAnalyzer normalizer,
+            NamedAnalyzer searchAnalyzer,
+            Builder builder
+        ) {
+            super(
+                name,
+                fieldType.indexOptions() != IndexOptions.NONE,
+                fieldType.stored(),
+                builder.hasDocValues.getValue(),
+                new TextSearchInfo(fieldType, builder.similarity.getValue(), searchAnalyzer, searchAnalyzer),
+                builder.meta.getValue()
+            );
+            setEagerGlobalOrdinals(builder.eagerGlobalOrdinals.getValue());
+            setIndexAnalyzer(normalizer);
+            setBoost(builder.boost.getValue());
+            this.ignoreAbove = builder.ignoreAbove.getValue();
+            this.nullValue = builder.nullValue.getValue();
+        }
+
+        public FlatObjectFieldType(String name, boolean isSearchable, boolean hasDocValues, Map<String, String> meta) {
+            super(name, isSearchable, false, hasDocValues, TextSearchInfo.SIMPLE_MATCH_ONLY, meta);
+            setIndexAnalyzer(Lucene.KEYWORD_ANALYZER);
+            this.ignoreAbove = Integer.MAX_VALUE;
+            this.nullValue = null;
+        }
+
+        public FlatObjectFieldType(String name) {
+            this(name, true, true, Collections.emptyMap());
+        }
+
+        public FlatObjectFieldType(String name, FieldType fieldType) {
+            super(
+                name,
+                fieldType.indexOptions() != IndexOptions.NONE,
+                false,
+                false,
+                new TextSearchInfo(fieldType, null, Lucene.KEYWORD_ANALYZER, Lucene.KEYWORD_ANALYZER),
+                Collections.emptyMap()
+            );
+            this.ignoreAbove = Integer.MAX_VALUE;
+            this.nullValue = null;
+        }
+
+        public FlatObjectFieldType(String name, NamedAnalyzer analyzer) {
+            super(name, true, false, true, new TextSearchInfo(Defaults.FIELD_TYPE, null, analyzer, analyzer), Collections.emptyMap());
+            this.ignoreAbove = Integer.MAX_VALUE;
+            this.nullValue = null;
+        }
```

Comment:
This is a lot of constructors with independent implementations. It's not clear when you would want to use each one. Is there a reasonable way to redirect all constructors into one implementation with `this(<params>)`. If not, could you add some comments to explain why each constructor exists?

Also, if this is a `public` class with a bunch of `public` methods, maybe it should be a top-level class in another file, rather than being an inner class.

Reply:
Oh... is this all copied from KeywordFieldType? 

That code is kind of ugly too, but I can see where it came from. 😛

Reply:
yes the FlatObjectFieldType is from KeywordFieldType, I was once thinking about if we can split the KeywordFieldType and then extend from it, but I realized all other fieldType are inner classes, so trying to follow the same way.